Water Thermal Tank Resulting in Negative Electric Consumption Fix

RKStrand opened the pull request:
Pull request overview
---------------------
 - Fixes #9001 
 - The water thermal tank model was generating negative electric consumption results that caused the entire building to report negative electric for a user's input file.  The problem was traced back to the fact that the user had requested autosizing of the maximum tank capacity for a heat pump water heater system.  In the code, the backup element capacity was set to the maximum tank capacity during the processing of input.  The problem though is that at that point the maximum tank capacity was still set to the autosizing parameter (-99999.0) which set the backup element capacity to -99999.  Then, during the run, when the backup element capacity was referenced and used, it set various parameters based on this negative value which in turn resulted in negative electric consumption.  The solution here was to not reset the backup element capacity if the maximum tank capacity has not yet been autosized or if the backup element capacity is set to something greater than or equal to zero.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [X] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

RKStrand committed:
Water Thermal Tank Negative Electric Fix

Water Thermal Tank was reporting negative electricity.  This was eventually traced back to the backup heating element being set before autosizing took place.  As a result, it was set to -99999 (DataSizing:AutoSize value) and this led to negative electric consumption.  This commit correct the problem by not allowing the backup element capacity to be set if autosizing has not yet been done.

RKStrand committed:
Merge branch 'develop' into 9001WaterThermalTankNegativeElectricityIssue

RKStrand committed:
Water Thermal Tank Negative Electric Unit Test

The solution to the original defect issue was converted into a small worker subroutine and a unit test was created to test the solution.

RKStrand committed:
9001 Correction of Unit Test

Bad cut and paste led to the new unit test crashing.  Not sure how that was not noticed previously but now it's fixed.

RKStrand committed:
Merge branch 'develop' into 9001WaterThermalTankNegativeElectricityIssue

rraustad commented on src/EnergyPlus/WaterThermalTanks.cc:
```diff
@@ -4228,7 +4228,7 @@ bool GetWaterThermalTankInput(EnergyPlusData &state)
 
                     // save backup element and on/off-cycle parasitic properties for use during standard rating procedure
-                    HPWH.BackupElementCapacity = Tank.MaxCapacity;
+                    setBackupElementCapacity(HPWH.BackupElementCapacity, Tank.MaxCapacity);
```
This is in GetWaterThermalTankInput. So if the tank max capacity is still equal to autosize here, which makes sense if sizing has not been called, then where is the backup element size set after GetWaterThermalTankInput is called? Is GetWaterThermalTankInput called more than once? If not, then there needs to be somewhere in init where the backup element is set using valid data? Since the comment mentions standard rating procedures, the backup element must be set before that call. Or another call to setBackupElementCapacity within the call to standard rating? since Tank.MaxCapacity would need to be known by then. Or right before the call to standard ratings inside somewhere that is a one time init.

rraustad replied:
Maybe this is handled in the sizing routine?

rraustad replied:
I see this same code in one other place. This same function used there as well?

    DesuperHtr.BackupElementCapacity = Tank.MaxCapacity;

rraustad replied:
How about setting the backup element capacity right here in sizing?

                if (this->MaxCapacityWasAutoSized) {
                    this->MaxCapacity = tmpMaxCapacity;
                    BaseSizer::reportSizerOutput(state, this->Type, this->Name, "Maximum Heater Capacity [W]", this->MaxCapacity);
                }

rraustad replied:
Setting the backup element capacity here probably means the new function is no longer needed.

RKStrand replied:
@rraustad I see what you are saying--setting the backup capacity in WaterThermalTankData::SizeStandAloneWaterHeater makes sense if it is autosized.  It would have to be done in several places in that routine since there are different sizing types for this unit.  I'll look into implementing this.

RKStrand replied:
But I still wonder if we don't want to set the backup capacity to -99999 just to be sure.  Maybe we don't need the new function.  In the case of a user inputed capacity, it can assign the backup capacity right away.  But if the backup capacity gets set to autosize, I feel like that's a bit dangerous potentially.  Maybe instead of a straight assignment, the code should say:

if (Tank.MaxCapacity < 0.0) HPWH.BackupElementCapacity = Tank.MaxCapacity;

If the BackupElementCapacity is initialized to zero, then it avoids the problematic situation where right now it gets set to -99999 and never recovers.
What do you think of that?  As in, get rid of the new function, modify the Get routine with the if from above and also for the Desuperheater, and then setting the backup capacity in the sizing routines (12 different places where this->MaxCapacity = tmpMaxCapacity happens).  Would that satisfy your concerns?  Biggest thing I haven't looked at yet is how to access the correct Backup Capacity.

RKStrand replied:
I think now that the new function should be for these new 12 assignment occurrences rather than how I have it set up now.  Since the Tank will need to reference a heat pump (this->HeatPumpNum) or a desuperheater (this->DesuperheaterNum), there would be good reason to not have to repeat this logic in all 12 places.  Then the unit test could be for this function not the one that I previously added.

rraustad replied:
tmpMaxCapacity is set multiple times, MaxCapacity is only set once at the end of all those assignments. Set backup element capacity there and that should be it.

rraustad replied:
There is this, if (this->MaxCapacityWasAutoSized), that lets us know that the MaxCapacity is autosized, there could also be one for backup element capacity but the backup is set equal to the MaxCapacity so it seems if MaxCapacity is autosized then they would still be equal after sizing. I am not positive on this but I did see several places where the max cap is reset for some reason. Worst case to me is if backup element capacity is autosized then set equal to MaxCapacity in sizing.

            // Reset the tank's internal heating element capacity.
            this->MaxCapacity = HeatPump.BackupElementCapacity;
            this->MinCapacity = HeatPump.BackupElementCapacity;

rraustad replied:
You could also use:

    if (BackupElementCapacity == DataSizing::AutoSize) {
        BackupElementCapacity = MaxCapacity;
    }

which would be valid if this is set in getInput.

    HPWH.BackupElementCapacity = Tank.MaxCapacity;

RKStrand replied:
If I remember correctly (juggling 5 different defects in the review process right now so it's possible that I am wrong here), there are three different sizing routines and at least two different types of equipment (heat pump and desuperheater) that use different data structures.  I'm not sure it will be super easy, but I think I get your point.  It probably won't be as complex (maybe 3 different places rather than greater than 12) as I thought.  I will try to work on this soon (hopefully later today?) and get a new commit made.

RKStrand committed:
Merge branch 'develop' into 9001WaterThermalTankNegativeElectricityIssue

RKStrand committed:
9001 Water Tank Negative Electric Review Comment Fixes

Based on review comments by @rraustad, rewrote the assignments that should avoid the negative backup element capacities that were generating negative electricity.

RKStrand committed:
Merge branch 'develop' into 9001WaterThermalTankNegativeElectricityIssue

RKStrand commented:
@rraustad Rewrote this based on your comments.  Much improved and I think it should capture all the potential bad assignments that are potentially possible rather than just the one case.  Revised unit test to match the new solution.  Hopefully CI comes back clean.

RKStrand committed:
9001 Water Thermal Tank fix weird typo

Title says it all.

rraustad commented:
2 files with warmup convergence diffs and very tiny eso diffs. But there is an unexpected change in coil sizing that needs explanation.

![image](https://user-images.githubusercontent.com/5778156/132097505-60053313-1a45-4026-9e5f-85e2ecd21839.png)

rraustad commented:
@Myoldmopar @mjwitte I am trying to run SingleFamilyHouse_HP_Slab locally to verify there are no diffs and the reason for these diffs is that this branch is behind develop. I ran ExpandObjects on the in.idf and then saved expanded.idf as in.idf and then get this error. I don't know if I have left out a step for conversion. I tried several different versions of ExpandObjects with the same result.

SingleFamilyHouse_HP_Slab.idf:

    Program Version,EnergyPlus, Version 9.6.0-47c7cd67bc, YMD=2021.09.04 17:00,
       ************* Beginning Zone Sizing Calculations
       ** Warning ** Weather file location will be used rather than entered (IDF) Location object.
       **   ~~~   ** ..Location object=FAIRBANKS INTL ARPT_AK_USA DESIGN_CONDITIONS
       **   ~~~   ** ..Weather File Location=Chicago Ohare Intl Ap IL USA TMY3 WMO#=725300
       **   ~~~   ** ..due to location differences, Latitude difference=[22.84] degrees, Longitude difference=[59.93] degrees.
       **   ~~~   ** ..Time Zone difference=[3.0] hour(s), Elevation difference=[51.13] percent, [68.00] meters.
       ** Severe  ** BuildingSurface:Detailed="FLOOR_UNIT1", invalid Outside Boundary Condition Object="SURFPROPOTHSDCOEFSLABAVERAGE".
       **   ~~~   **  no OtherSideCoefficients of that name.

From expanded.idf:

    !   BuildingSurface:Detailed,
    !     Floor_unit1,             !- Name
    !     Floor,                   !- Surface Type
    !     Interior Floor,          !- Construction Name
    !     living_unit1,            !- Zone Name
    !     ,                        !- Space Name
    !     GroundSlabPreprocessorAverage,  !- Outside Boundary Condition
    !     surfPropOthSdCoefSlabAverage,  !- Outside Boundary Condition Object
    !     NoSun,                   !- Sun Exposure

    BuildingSurface:Detailed,
      Floor_unit1,                                             !- Name
      Floor,                                                   !- Surface Type
      Interior Floor,                                          !- Construction Name
      living_unit1,                                            !- Zone Name
      ,                                                        !- Space Name
      OtherSideCoefficients,                                   !- Outside Boundary Condition
      surfPropOthSdCoefSlabAverage,                            !- Outside Boundary Condition Object
      NoSun,                                                   !- Sun Exposure

IDD:

    BuildingSurface:Detailed,
      A6 , \field Outside Boundary Condition
           \required-field
           \type choice
           \key Adiabatic
           \key Surface
           \key Zone
           \key Outdoors
           \key Foundation
           \key Ground
           \key GroundFCfactorMethod
           \key OtherSideCoefficients
           \key OtherSideConditionsModel
           \key GroundSlabPreprocessorAverage
           \key GroundSlabPreprocessorCore
           \key GroundSlabPreprocessorPerimeter
           \key GroundBasementPreprocessorAverageWall
           \key GroundBasementPreprocessorAverageFloor
           \key GroundBasementPreprocessorUpperWall
           \key GroundBasementPreprocessorLowerWall

RKStrand commented:
@Myoldmopar @mjwitte I am getting a similar problem but not exactly the same on the Mac.  Though I am using the V9.5 ExpandObjects.  Here is the idf object that is problematic:

BuildingSurface:Detailed,
  Floor_unit1,                                             !- Name
  Floor,                                                   !- Surface Type
  Interior Floor,                                          !- Construction Name
  living_unit1,                                            !- Zone Name
  ,                                                        !- Outside Boundary Condition
  **GroundSlabPreprocessorAverage**,                           !- Outside Boundary Condition Object
  surfPropOthSdCoefSlabAverage,                            !- Sun Exposure
  NoSun,                                                   !- Wind Exposure
  NoWind,                                                  !- View Factor to Ground
  0.0,                                                     !- Number of Vertices
  4,                                                       !- Vertex 1 X-coordinate
  0,                                                       !- Vertex 1 Y-coordinate
  0,                                                       !- Vertex 1 Z-coordinate
  0.0101,                                                  !- Vertex 2 X-coordinate
  0,                                                       !- Vertex 2 Y-coordinate
  9.09981820971244,                                        !- Vertex 2 Z-coordinate
  0.0101,                                                  !- Vertex 3 X-coordinate
  12.1330909462833,                                        !- Vertex 3 Y-coordinate
  9.09981820971244,                                        !- Vertex 3 Z-coordinate
  0.0101,                                                  !- Vertex 4 X-coordinate
  12.1330909462833,                                        !- Vertex 4 Y-coordinate
  0,                                                       !- Vertex 4 Z-coordinate
  0.0101;                                                  !- Vertex 5 X-coordinate

I get a slightly different error message (that I need to run ExpandObjects) even though I did actually run it.
When I fix the outside boundary condition object type to OtherSideCoefficients, I get the same error message as @rraustad.  There are simply no other side coefficients in the idf.

RKStrand commented:
@rraustad The two files that are flagged in the ci results are the only files in the test suite that have a wrapped condenser and autosize the tank heater capacity (based on a search of the idfs).  Those files use a stratified tank rather than a mixed tank.  So, I'm going to take a look and see if I can figure out why that is happening based on that information.

RKStrand commented:
@rraustad Would a wrapped condenser have a backup heating element?  Since the only thing that has changed here presumably is the setting of the backup heating element, should the routine that is setting the backup heating element skip wrapped condensers if that is even possible?

rraustad commented:
@RKStrand I would think that all HP water heaters would have a backup element in the tank. Whether that was accounted for during model development is another question.

mjwitte commented:
@RKStrand @rraustad Any file that uses a `Ground*Preprocessor` boundary condition requires the latest expandobjects (due to the new Space Name field in surfaces and the reduced fields in AirloopHVAC:OutdoorAirSystem) and it must be run using EP-Launch or RunEPlus.bat, because (if I remember correctly) the command line option to run expandobjects does not support the extra steps required for the Slab and Basement preoprocessors. The OtherSideCoefficients error indicates that the file copy from the slab output into the idf didn't happen.

RKStrand committed:
Merge branch 'develop' into 9001WaterThermalTankNegativeElectricityIssue

RKStrand committed:
Merge branch 'develop' into 9001WaterThermalTankNegativeElectricityIssue

RKStrand committed:
9001 Trial to Check Autosized DX Coil

During CI testing for the defect fix, a difference in the coil sizing for a wrapped condenser was noted.  This backs out the change for that specific type of system to see if this is the cause.

RKStrand commented:
@rraustad I'm testing out a new commit to see if it makes any difference with respect to the differences noted in the wrapped condenser files (two SingleFamilyHome idfs).  I hope CI will come back clean other than the two warmup differences that don't appear to be a big deal.  If it does, we can decide how to proceed.

RKStrand commented:
@rraustad I have made a change that gets rid of all the differences between this branch and develop.  Basically, it doesn't do anything if it is a wrapped condenser while still fixing the problem for all the other cases.  I think there are a couple of options here:

1. Keep this latest variation and call it good (because it fixes the problem and does not affect anything else).  In the SingleFamilyHouse file that I'm using right now, the backup capacity never gets reset from -99999 but clearly that doesn't bother the simulation since it runs to completion and doesn't report any negative electric values.
2. Get rid of the new variation that skips wrapped condensers and modify the new subroutine (setBackupElementCapacity) so that if the MaxCapacity is zero or less that the BackupElementCapacity gets set to zero.  I'm not sure what the outcome will be here and what it might mess up.
3. Get rid of the new variation that skips wrapped condensers and the call it good.  This still fixes the problem and the only differences are a roughly 8W difference in sizing.  Basically we are then saying this is fine or "more correct".
4. Some other iteration I haven't thought of yet.

What do you think about these options?  With the upcoming code freeze, I'd like to get this resolved so this can get merged in before the freeze.  Thanks!

rraustad commented:
Well, I know I saw somewhere that the tank capacity got reset to backup capacity. Maybe that doesn't happen for the wrapped condenser? Is there any time when a backup element would be needed for a wrapped condenser? Low ambient, etc. Is the backup element capacity reported to the eio? And if so is it -9999.0? SingleFamilyHouse uses Fairbanks AK as location and runs Chicago Ohare on the CI. I would think this would trip a low ambient. I would like to know for sure if the backup element should every be used for the wrapped condenser. What if you increase the draw on the tank without increasing capacity for that example file? What does the tank exiting temperature profile (min temp) look like on an annual run.

Down in this function is this. Does this get hit for the wrapped condenser, and if so what is the tank max capacity.

    WaterThermalTankData::CalcStandardRatings

                //       reset the water heater data to original values
                this->MaxCapacity = state.dataWaterThermalTanks->HPWaterHeater(HPNum).BackupElementCapacity;

RKStrand commented:
@rraustad Ok, a little update here.  I've been tracking BackupElementCapacity in the latest and the previous version.  What I am seeing is that in the latest version (same as develop for wrapped condenser) is that BackupElementCapacity never changes from -99999.  In the previous version (that had the difference in the coil sizing), the value of BackupElementCapacity gets changed to 5500 which is what the HP gets sized to for heating.  So, based on that, I think the previous edition is actually the right solution and I think the sizing difference is legitimate since there is finally a BackupElementCapacity that is not -99999.  The specific line you mention in CalcStandardRatings--I put a breakpoint there and it never gets to that point.  So, it seems that once the heat pump is sized, the previous version does pick that up like it should.  I'm more convinced now that the "previous" version with the small changes in the coil sizing results is actually correct.  Does this help at all or just make things less clear?

rraustad commented:
That makes sense. Move the break point to the start of the StandardRating function and see where it goes.

RKStrand commented:
@rraustad Well, it goes through the standard rating process and then resets BackupElementCapacity back to the MaxCapacity at the end which is -99999 for both versions.  However, in the previous version, it resets the BackupElementCapacity to 5500 somewhere else.  I put a breakpoint on every occurrence of BackupElementCapacity getting set to something.  What's vexing is I don't see where that happens leading me to think that there is some other weird assignment thing happening.

RKStrand commented:
@rraustad Update...the new subroutine is the place where it gets updated.  I got tired of hitting continue numerous times and deactivated the breakpoint.  This time I kept hitting continue and eventually it got here.  In develop, it would never have updated this value.  The new subroutine (previous commit edition) appears to be doing what it is supposed to do--fix BackupElementCapacity once MaxCapacity gets reset.  So, I think this is what we want and should be more correct than develop even though we are now getting a slight difference in the coil size.  I think maybe the coil sizing routine must sense the there is something there whereas before it was ignoring a clearly wrong negative number?  So, I feel fairly confident that the previous commit is the one we want because it fixes the negative electric and actually sets the BackupElementCapacity correctly.  Develop fails on both of those criteria and the latest commit fails on setting the BackupElementCapacity correctly.  Do you agree with that assessment?  If so, I will back out the changes from the latest commit.  We will have those small diffs in the two files but I think that's okay because it should be more correct.  If not, what do you feel like will push this over the line?

rraustad commented:
Maybe MaxCapacity is getting reset to something else after leaving StandardRatings and before tank calculations are performed. It just seems that the backup element should not be left at the autosize state after sizing. I was more comfortable with your previous fix it's just that the diffs had to be explained. If backup element sizing was the only difference, then a change in sizing must have been due to correcting the backup element size. Looking at the 2 files with diffs from 2 days ago, it's only the SingleFamilyHouse files. And only big table diffs. Maybe you did have it correct the second time. The coil sizing report shows the HP evaporator coil only changed by 27 W but I don't see a change in air flow or capacity in the eio file so where is this diff coming from? It is almost like this branch was not current with develop.

![image](https://user-images.githubusercontent.com/5778156/133493057-6677a15d-07a0-4d72-a4b0-e46edde7e64a.png)

rraustad commented:
I usually run the files locally in develop and my branch using the weather used by the CI, for these 2 files it's chicago. Then open both html's and check for diffs in this same place.

add_simulation_test(IDF_FILE SingleFamilyHouse_HP_Slab.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
add_simulation_test(IDF_FILE SingleFamilyHouse_HP_Slab_Dehumidification.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)

RKStrand commented:
@rraustad So, it looks like the capacity for the evaporator are not listed in the eio.  However, if you look at the idf, the Heat Pump Water Heater Evaporator_unit1 is NOT autosized.  It's a hard wired value of 1500W.  So, why would it be "sized" to something different.  The place where there is an autosize is in the water tank heater capacity in WaterHeater:Stratified. So, if there was a difference, I would have thought that the heater capacity for the tank would be different.  But there is no difference shown in the eio.  (By the way, the chart you posted has units of Btu/h because the idf requests IP output units.)

RKStrand committed:
Merge branch 'develop' into 9001WaterThermalTankNegativeElectricityIssue

rraustad commented on src/EnergyPlus/WaterThermalTanks.cc:
```diff
@@ -12603,6 +12609,23 @@ void WaterThermalTankData::oneTimeInit(EnergyPlusData &state)
+        }
+    }
+}
+
```
@RKStrand I posted this comment here since I thought updating the HP and desuperheater backup element capacity may not be quite correct since the backup resistive element is in the tank. Writing out this comment and searching through code makes me think this is the correct change.

I pulled this branch to review. I was looking for a backup element and that input does not exist. It seems it is an inferred parameter for the models that interact with tanks. But it is not even used in the tank model (or maybe it does and I can't see it on cursory review). It is used in the models that interact with the tank. The HP and desuperheater pull this back-door variable from the tanks. This review shows:

1) there are no diffs now.
2) When these other models pull the tank max capacity into a backup element variable, and that max capacity was autosized, this change now updates those values correctly (after the tank capacity is autosized) from -9999 to whatever the tank capacity sizes to. I agree with what you did here.
3) If this backup element is used for something within a model, the corner case of its use is not exercised in an example file, otherwise there would be eio diffs (hard-sized coil capacities do not report to the eio, and changes in tank temp would result in plant diffs).

![image](https://user-images.githubusercontent.com/5778156/133710383-c7012f11-398f-4589-b061-d01caed8ee5a.png)

The mixed tank has a single max capacity. This model, I would think, would have a backup element of the same capacity. The stratified tank has 2 capacity inputs (resistive elements). That model should NOT have a backup element (backup capacity = 0 since there are already 2 elements). So without rethinking the model intent I have to agree with what you have done for this specific issue.

    WaterHeater:Mixed,
      N4 , \field Heater Maximum Capacity

    WaterHeater:Stratified,
      N6 , \field Heater 1 Capacity
      N9 , \field Heater 2 Capacity

rraustad commented:
@RKStrand I think this is done. Respond otherwise if you think differently (e.g., example file updates, model improvements, etc, posted as new issues). Otherwise this should merge by end of day tomorrow. One thought just crossed my mind, how do we even know if the backup element variable used in these models is ever activated during a specific time step? Is there any form of report when this happens? If not, another good follow up.

rraustad commented:
I have an alternate suspicion that this backup element variable may be nothing more than a storage location for the original max capacity for times when the HP and desuperheater modify the real time capacity and that value needs to get reset for the next time step. That would actually make more sense but I am not sure since we all know there are 2 elements in a HW tank. Kinda supports good comments in code. This thought does not change my approval of this branch.

rraustad commented:
The declaration does say it is a backup (independent) element. So back to merge this and move on.

        Real64 BackupElementCapacity;     // Tank backup element capacity (W)

RKStrand commented:
@rraustad Thanks for looking back at this PR again!  I agree with you--what we have fixes the initial problem and is clearly needed to avoid the divide by zero.  From my standpoint, I think this is good to go.  I looked a lot more at this yesterday stepping through debugger for the SingleFamilyHome input file and can't find a compelling reason as to why this solution should not be done.  I am not sure whether this is a separate element for heating or simply something that is storing the actual heater capacity so it is not lost.  I had not even thought of that second option--one that actually makes some sense, but the code is not 100% clear on that.  In either case though, I think this is good to go, and we are getting no diffs in the results.  So, I think this is ready to merge by whoever has merging privileges.

Myoldmopar commented:
Looks like @rraustad and @RKStrand are both satisfied with this.  Maybe?

> merge this and move on.
       - @rraustad

> I think this is ready to merge
       - @RKStrand 

I'm going to pull the branch and build with latest develop and see what happens.  I'll review the code changes while that is happening.  If anyone thinks this isn't ready, speak up...

Myoldmopar commented:
Pulled develop and built no problem.  Ran tests, also no problem.  Took defect file and weather data and ran them as well.  I didn't see any negative electricity values in the tabular output, but I'm not sure that is exactly verification of the fix.  I'm relying a bit on @RKStrand and @rraustad on that.  OK, merging this in.  Thanks!
